src/inmanta/ast/constraint/expression.py

```diff
@@ -236,34 +236,36 @@ def _validate_value(self, value: object, side: int) -> None:
             )
             raise e
 
-    def resume(self, requires, resolver, queue, target):
+    def resume(self, requires: Dict[object, object], resolver: Resolver, queue: QueueScheduler, target: ResultVariable) -> None:
         result = self.children[0].execute(requires, resolver, queue)
         self._validate_value(result, 0)
+        assert isinstance(result, bool)
         if self._is_final(result):
             target.set_value(result, self.location)
         else:
             ExecutionUnit(
                 queue, resolver, target, self.children[1].requires_emit(resolver, queue), self.children[1], owner=self
             )
 
-    def execute_direct(self, requires):
+    def execute_direct(self, requires: Dict[object, object]) -> object:
         lhs = self.children[0].execute_direct(requires)
         self._validate_value(lhs, 0)
+        assert isinstance(lhs, bool)
```

Comment:
I think we could get around the need for this if we would make `Operator` a generic in its return type. `_bin_op` and `execute` would then return `T` and `LazyBooleanOperator` would be `BinaryOperator[bool]`.

Reply:
Ah nevermind, `execute` gets overridden here to make it lazy, that complicates things.